File: solutions/ruby/binary-search-tree/1/binary_search_tree.rb
=begin
Write your code for the 'Binary Search Tree' exercise in this file. Make the tests in
`binary_search_tree_test.rb` pass.


To get started with TDD, see the `README.md` file in your
`ruby/binary-search-tree` directory.
=end
class Bst
  def initialize(data)
    @data = data
    @left = nil
    @right = nil
  end
  def data
    return @data
  end
  def left
    return @left
  end
  def right
    return @right
  end
  def insert(data)
    if data <= @data then
      if @left == nil then
        @left = Bst.new data
      else
        @left.insert data
      end
    else
      if @right == nil then
        @right = Bst.new data
      else
        @right.insert data
      end
    end
  end
  def each(&block)
    return enum_for(__method__) if block.nil?
    if @left != nil then
      @left.each do |value|
        block.call value
      end
    end
    block.call @data
    if @right != nil then
      @right.each do |value|
        block.call value
      end
    end
  end
end

File: solutions/ocaml/binary-search-tree/1/binary_search_tree.ml
open Base

type bst = Leaf | Node of {value : int; left : bst; right : bst}

let empty = Leaf

let value = function
  | Leaf -> Error "empty tree"
  | Node {value; _} -> Ok value

let left = function
  | Leaf -> Error "no"
  | Node {left; _} -> Ok left

let right = function
  | Leaf -> Error "no"
  | Node {right; _} -> Ok right

let rec insert x = function
  | Leaf -> Node {value = x; left = Leaf; right = Leaf}
  | Node {value; left; right} when x <= value -> Node {value; left = insert x left; right}
  | Node {value; left; right} when x > value -> Node {value; left; right = insert x right}

let rec to_list = function
  | Leaf -> []
  | Node {value; left; right} -> to_list left @ [value] @ to_list right

